pkg/engine/executor/evaluator.go

```diff
@@ -0,0 +1,111 @@
+package executor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"iter"
+
+	"github.com/apache/arrow-go/v18/arrow/memory"
```

Comment:
I've mentioned it to @chaudum in a 1:1: I'm a little critical regarding the Apache Arrow and its Go implementation. We've done a [hackathon](https://github.com/grafana/loki/tree/hackathon/hackathon-2023-12-arrow-engine) with it and came to two conclusions

1. The implementation and especially its SIMD based compute kernels weren't ready (that was end of 2023).
2. Loki's main bottlenecks are string processing and allocations. Numeric vectorization that Arrow etc bring have little impact. I do not know how well the Arrow is with strings and allocations, though.

That sad I do not want to impose and am pretty sure you've considered more options that I have. I just wanted to share my inside :slightly_smiling_face: